File: EECE7205-Advanced-Algorithms/HW4/dijkstra.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <map>
#include <climits>
#include <utility>
using namespace std;

vector<int> dijkstra(int src, int V, map<int, vector<pair<int, int>>>& G) {
    vector<int> d(V, INT_MAX);
    vector<bool> visited(V, false);
    vector<pair<int, int>> Q;

    d[src] = 0;
    Q.push_back({0, src});

    while (!Q.empty()) {
        int min_idx = 0;
        for (int i = 1; i < Q.size(); i++) {
            if (Q[i].first < Q[min_idx].first) {
                min_idx = i;
            }
        }

        int dist = Q[min_idx].first;
        int u = Q[min_idx].second;
        Q.erase(Q.begin() + min_idx);

        if (visited[u]) {
            continue;
        }
        visited[u] = true;

        if (G.find(u) != G.end()) {
            for (auto& e : G[u]) {
                int v = e.first;
                int w = e.second;

                if (!visited[v] && d[u] + w < d[v]) {
                    d[v] = d[u] + w;
                    Q.push_back({d[v], v});
                }
            }
        }
    }
    return d;
}

int main() {
    int V = 8;
    map<int, vector<pair<int, int>>> G;
    G[0] = {{1, 3}, {3, 7}};
    G[1] = {{2, 1}, {3, 4}};
    G[2] = {{3, 2}, {4, 5}};
    G[3] = {{4, 1}};
    G[4] = {{5, 7}, {6, 3}};
    G[5] = {{6, 2}, {7, 4}};
    G[6] = {{7, 6}};
    G[7] = {};

    int src = 0;
    vector<int> distances = dijkstra(src, V, G);

    cout << "Vertex\tDistance from Source\n";
    for (int i = 0; i < distances.size(); i++) {
        cout << i << "\t" << distances[i] << "\n";
    }
    return 0;
}

File: EECE7205-Advanced-Algorithms/HW4/dijkstra.py
import heapq
import networkx as nx
import matplotlib.pyplot as plt

def dijkstra(src, V, G_adj_list):
    Q = []
    d = [float('inf')] * V

    d[src] = 0
    heapq.heappush(Q, (0, src))

    while Q:
        dist, u = heapq.heappop(Q)

        if dist > d[u]:
            continue

        for v, w in G_adj_list[u]:
            if d[u] + w < d[v]:
                d[v] = d[u] + w
                heapq.heappush(Q, (d[v], v))

    print("Vertex\tDistance from Source")
    for vertex, distance in enumerate(d):
        print(f"{vertex}\t{'INF' if distance == float('inf') else distance}")
    
    return d

V = 8
G_adj_list = {
    0: [(1, 3), (3, 7)],
    1: [(2, 1), (3, 4)],
    2: [(3, 2), (4, 5)],
    3: [(4, 1)],
    4: [(5, 7), (6, 3)],
    5: [(6, 2), (7, 4)],
    6: [(7, 6)],
    7: [],
}

src = 0
distances = dijkstra(src, V, G_adj_list)

G = nx.DiGraph()
for u, neighbors in G_adj_list.items():
    for v, w in neighbors:
        G.add_edge(u, v, weight=w)

pos = nx.spring_layout(G)

plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
nx.draw(G, pos, with_labels=True, node_color='lightblue', node_size=500, font_size=10)
edge_labels = nx.get_edge_attributes(G, 'weight')
nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels)
plt.title("Original Graph")

SP_tree = nx.DiGraph()
for u in range(V):
    for v, w in G_adj_list[u]:
        if distances[v] == distances[u] + w:
            SP_tree.add_edge(u, v, weight=w)

plt.subplot(1, 2, 2)
nx.draw(SP_tree, pos, with_labels=True, node_color='lightgreen', node_size=500, font_size=10)
sp_edge_labels = nx.get_edge_attributes(SP_tree, 'weight')
nx.draw_networkx_edge_labels(SP_tree, pos, edge_labels=sp_edge_labels)
plt.title("Shortest Path Tree (SPT from Source)")

plt.tight_layout()
plt.show()
